system/inc/simple_pool_allocator.h

```diff
@@ -0,0 +1,159 @@
+                }
+                if (candidate == freeListEnd_) {
+                    freeListEnd_ = prev;
+                }
+                candidate->next = nullptr;
+                p = reinterpret_cast<void*>(candidate->data);
+            }
+        }
+        return p;
+    }
+
+    void deallocate(void* p) {
+        if (p == nullptr) {
+            return;
+        }
+        BlockHeader* block = reinterpret_cast<BlockHeader*>(reinterpret_cast<uint8_t*>(p) - sizeof(BlockHeader));
+        if ((reinterpret_cast<uint8_t*>(block) + block->size) == ptr_) {
+            ptr_ -= block->size;
+        } else {
+            BlockHeader* ins = nullptr;
+            for(BlockHeader* b = freeList_; b != nullptr; b = b->next) {
+                if (block > b) {
+                    ins = b;
+                }
+            }
+            if (ins == nullptr) {
+                if (freeList_ == nullptr) {
+                    freeListEnd_ = block;
+                }
+                block->next = freeList_;
+                freeList_ = block;
+            } else {
+                block->next = ins->next;
+                ins->next = block;
+                if (ins == freeListEnd_) {
+                    freeListEnd_ = block;
+                }
+            }
+        }
+        shrink();
+    }
+
+    virtual ~SimpleBasePool() = default;
+
+protected:
+    SimpleBasePool(void* location, size_t size) :
+        begin_(reinterpret_cast<uint8_t*>(location)),
+        size_(size),
+        ptr_(begin_) {
+    }
+
+    uint8_t* begin_;
+    size_t size_;
+
+// Leaving these protected to simplify unit testing
+// private:
+
+    struct BlockHeader {
+        BlockHeader* next;
+        uintptr_t size;
+        uint8_t data[0];
+    };
+
+    static_assert(sizeof(BlockHeader) % sizeof(uintptr_t) == 0, "SimpleBasePool: size of header should be a multiple of uintptr_t");
+
+    static size_t aligned(size_t sz) {
+        return (sz + (sizeof(uintptr_t) - (sz % sizeof(uintptr_t))));
+    }
+
+    BlockHeader* getPrevFree(BlockHeader* block) const {
+        for (BlockHeader* b = freeList_; b != nullptr; b = b->next) {
+            if (b->next == block) {
+                return b;
+            }
+        }
+        return nullptr;
+    }
+
+    void shrink() {
+        while (freeListEnd_ != nullptr && (reinterpret_cast<uint8_t*>(freeListEnd_) + freeListEnd_->size) == ptr_) {
+            ptr_ -= freeListEnd_->size;
+            freeListEnd_ = getPrevFree(freeListEnd_);
+            if (freeListEnd_ == nullptr) {
+                freeList_ = nullptr;
+            } else {
+                freeListEnd_->next = nullptr;
+            }
+        }
+    }
+
+    uint8_t* ptr_;
+
+    BlockHeader* freeList_ = nullptr;
+    BlockHeader* freeListEnd_ = nullptr;
+};
+
+class SimpleAllocedPool : public SimpleBasePool {
+public:
+    SimpleAllocedPool(size_t size) :
+        SimpleBasePool(reinterpret_cast<void*>(new uint8_t[size]), size) {
```

Comment:
I think `SimpleAllocedPool` (and its base class) should be tolerant to memory allocation errors, given that it can be requested to allocate a relatively large buffer and that the exact size of the heap depends on a user application.

Reply:
👍